File: 223.rectangle-area.py
#
# @lc app=leetcode id=223 lang=python3
#
# [223] Rectangle Area
#

# Time:  O(1)
# Space: O(1)

# @lc code=start
class Solution:
    def computeArea(self, A: int, B: int, C: int, D: int, E: int, F: int, G: int, H: int) -> int:
        sum_area = (C - A) * (D - B) + (G - E) * (H - F)
        if A >= G or B >= H or C <= E or D <= F:
            return sum_area
        dup_area = (min(C, G) - max(A, E)) * (min(D, H) - max(B, F))
        return sum_area - dup_area


# @lc code=end

File: 223.rectangle-area.js
/*
 * @lc app=leetcode id=223 lang=javascript
 *
 * [223] Rectangle Area
 */

// @lc code=start
/**
 * @param {number} A
 * @param {number} B
 * @param {number} C
 * @param {number} D
 * @param {number} E
 * @param {number} F
 * @param {number} G
 * @param {number} H
 * @return {number}
 */
var computeArea = function (A, B, C, D, E, F, G, H) {
  const sumArea = (C - A) * (D - B) + (G - E) * (H - F);
  if (A >= G || B >= H || C <= E || D <= F) {
    return sumArea;
  }
  const dupX = Math.min(C, G) - Math.max(A, E);
  const dupY = Math.min(D, H) - Math.max(B, F);
  return sumArea - dupX * dupY;
};
// @lc code=end
